### ISSUE TITLE
Class unavailable to ClassLoader crashes JavaAssist

### ISSUE THREAD

**Issue by cmelchior** [jboss-javassist/javassist#70]
**Use case**: An Android library with optional dependencies provide some methods that are only callable if the user provide the dependency themselves. While a bit unconventional it helps to save both file size and methods which are important on Android. 

Example:
```
public class Foo {
  // Methods

 // Method requiring RxJava as dependency. If people never call this method, they don't have to include RxJava
 public rx.Observable asObservable() {
   // returns an RxJava observable
 }
```

Normally users don't encounter problems as anyone trying to call the `asObservable` method from their IDE will get compile time warnings and they can add the dependency.

**Problem:**
However JavaAssist is used by tools like PowerMock when writing unit tests. JavaAssist unfortunately seem to throw a hard exception if it encounters a class it cannot load:

```
Caused by: javassist.NotFoundException: rx.Observable
	at javassist.ClassPool.get(ClassPool.java:450)
	at javassist.bytecode.Descriptor.toCtClass(Descriptor.java:592)
	at javassist.bytecode.Descriptor.getReturnType(Descriptor.java:489)
	at javassist.CtBehavior.getReturnType0(CtBehavior.java:306)
	at javassist.CtMethod.getReturnType(CtMethod.java:219)
```

I am a bit uncertain if this is something that should be solved at the JavaAssist level or if it is a library problem (e.g. PowerMock needs to handle it).

**Comment by chibash** [comment]
The stack trace seems to show that PowerMock tried to get the return type of asObservable().

**Comment by cmelchior** [comment]
Yes, and that is the problem. The return type of `asObservable()` is a `rx.Observable()` which isn't available to the class loader.

**Comment by chibash** [comment]
So, if PowerMock requires the return type of `asObservable()`, Javassist could not fix the problem.
It could be possible to postpone reading a class file but, if PowerMock needs to check its definition, the class file has to be loaded anyway and an exception will be thrown.

**Comment by cmelchior** [comment]
Yes, as I wrote in the original post I was also a bit unsure if this is JavaAssists problem. Going over the code again ( https://github.com/jboss-javassist/javassist/blob/master/src/main/javassist/ClassPool.java#L444-Lundefined ) I noticed that `NotFoundException` is actually a checked exception, which is probably the best JavaAssist can do since any libraries using JavaAssist should then be aware that this issue could occur.

I think you are right, JavaAssist is doing the best it can in this situation and it needs to be handled higher up in the stack. Sorry for the trouble.
